integration-tests/tests/src/tests/notifications.ts

```diff
@@ -0,0 +1,463 @@
+      const handler = () => {
+        const objects = this.realm.objects(TestObject);
+        runCount++;
+        if (runCount === 1) {
+          expect(objects.length).equals(1);
+          expect(objects[0].doubleCol).equals(1);
+        } else {
+          throw new Error("Should not run after removeListener");
+        }
+      };
+
+      this.realm.addListener("change", handler);
+
+      this.realm.write(() => {
+        this.realm.create("TestObject", { doubleCol: 1 });
+      });
+
+      this.realm.removeListener("change", handler);
+
+      this.realm.write(() => {
+        this.realm.create("TestObject", { doubleCol: 2 });
+      });
+
+      expect(runCount).equal(1);
+    });
+
+    it("implements removeAllListeners", function (this: RealmContext) {
+      const handler = () => {
+        const objects = this.realm.objects(TestObject);
+        runCount++;
+        if (runCount === 1) {
+          expect(objects.length).equals(1);
+          expect(objects[0].doubleCol).equals(1);
+        } else {
+          throw new Error("Should not run after removeListener");
+        }
+      };
+
+      const secondHandler = () => {
+        const objects = this.realm.objects(TestObject);
+        runCount++;
+        if (runCount === 2) {
+          expect(objects.length).equals(1);
+          expect(objects[0].doubleCol).equals(1);
+        } else {
+          throw new Error("Should not run after removeAllListeners");
+        }
+      };
+
+      this.realm.addListener("change", handler);
+      this.realm.addListener("change", secondHandler);
+
+      this.realm.write(() => {
+        this.realm.create("TestObject", { doubleCol: 1 });
+      });
+
+      this.realm.removeAllListeners("change");
+
+      this.realm.write(() => {
+        this.realm.create("TestObject", { doubleCol: 2 });
+      });
+
+      expect(runCount).equal(2);
+    });
+  });
+
+  /**
+   * Helper function to use as a generic handler for collection listeners.
+   * @param onEnd Callback to run once the listener reaches the end of the changes array. If the callback is Mocha.Done, it will ensure that
+   * the listener was run exactly [changesOnRun.length] times
+   * @param changesOnRun An array of Realm.CollectionChangeSet.
+   * The handler will assert equality of listener's changes with the next element of the array on every run.
+   */
+  const expectCollectionChangesOnEveryRun = (
+    onEnd: () => void | Mocha.Done,
+    changesOnRun: Realm.CollectionChangeSet[],
+  ) => {
+    return (_: Realm.Collection<Realm.Object>, changes: Realm.CollectionChangeSet) => {
+      runCount++;
+
+      expect(changes.insertions).deep.equals(changesOnRun[runCount - 1].insertions);
+      expect(changes.oldModifications).deep.equals(changesOnRun[runCount - 1].oldModifications);
+      expect(changes.newModifications).deep.equals(changesOnRun[runCount - 1].newModifications);
+      // expect(changes.modifications).deep.equals(changesOnRun[runCount - 1].modifications);
```

Comment:
Should I run an assert on modifications? I could also compare it with the new/oldModifications, whichever was the one this actually represents.